camel/messages/base.py

```diff
@@ -271,6 +281,108 @@ def extract_text_and_code_prompts(
 
         return text_prompts, code_prompts
 
+    @classmethod
+    def from_sharegpt(
+        cls,
+        message: ShareGPTMessage,
+        function_format: Optional[FunctionCallFormatter[Any, Any]] = None,
+        role_mapping=None,
+    ) -> "BaseMessage":
+        """Convert ShareGPT message to BaseMessage or FunctionCallingMessage.
+        Note tool calls and responses have an 'assistant' role in CAMEL
+
+        Args:
+            message (ShareGPTMessage): ShareGPT message to convert
+            function_format (FunctionCallFormatter, optional):
+             Function call formatter to use. Defaults to Hermes.
+            role_mapping (Dict[str, List[str, RoleType]], optional):
+             Role mapping to use. Defaults to a CAMEL specific mapping.
+
+        Returns:
+            BaseMessage: Converted message
+        """
+
+        if role_mapping is None:
+            role_mapping = {
+                "system": ["system", RoleType.USER],
+                "human": ["user", RoleType.USER],
+                "gpt": ["assistant", RoleType.ASSISTANT],
+                "tool": ["assistant", RoleType.ASSISTANT],
+            }
+        role_name, role_type = role_mapping[message.from_]
+
+        if function_format is None:
+            function_format = HermesFunctionFormatter()
+
+        # Check if this is a function-related message
+        if message.from_ == "gpt":
+            func_info = function_format.extract_tool_calls(message.value)
+            if (
+                func_info and len(func_info) == 1
+            ):  # TODO: Handle multiple tool calls
+                # Including cleaned content is useful to
+                # remind consumers of non-considered content
+                clean_content = re.sub(
+                    r"<tool_call>.*?</tool_call>",
+                    "",
+                    message.value,
+                    flags=re.DOTALL,
+                ).strip()
+                from camel.messages import FunctionCallingMessage
+
+                return FunctionCallingMessage(
+                    role_name=role_name,
+                    role_type=role_type,
+                    meta_dict=None,
+                    content=clean_content,
+                    func_name=func_info[0].__dict__["name"],
+                    args=func_info[0].__dict__["arguments"],
+                )
```

Comment:
since ShareGPT data will be structured, would it be simpler and more reliable to convert the JSON data directly into a dictionary and then extract the relevant information from this structured format, rather than using `extract_tool_calls`, which uses regex?

Reply:
My understanding is ShareGPT doesn't actually say anything about function calling, that depends on the function calling format which may be part of a ShareGPT formatted conversation. In the case of Hermes for example, its messages can be in a ShareGPT conversation, and its function calls are located by and separated by <tool_call> xml tags (relevant when we support multiple tool calls per message). It will also be important to remove the function call parts of an assistant message when only part of the assistant message is meant to be e.g. shown to a user